zcash_client_sqlite/src/lib.rs

```diff
@@ -1049,19 +1050,39 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
     ) -> Result<(), Self::Error> {
         self.transactionally(|wdb| {
             let tx_ref = wallet::put_tx_data(wdb.conn.0, d_tx.tx(), None, None)?;
+            let funding_accounts = wallet::get_funding_accounts(wdb.conn.0, d_tx.tx())?;
+            let funding_account = funding_accounts.iter().next().copied();
+            if funding_accounts.len() > 1 {
+                warn!(
+                    "More than one wallet account detected as funding transaction {:?}, selecting {:?}",
+                    d_tx.tx().txid(),
+                    funding_account.unwrap()
+                )
+            }
 
-            let mut spending_account_id: Option<AccountId> = None;
             for output in d_tx.sapling_outputs() {
                 match output.transfer_type() {
-                    TransferType::Outgoing | TransferType::WalletInternal => {
-                        let recipient = if output.transfer_type() == TransferType::Outgoing {
-                            //TODO: Recover the UA, if possible.
-                            Recipient::Sapling(output.note().recipient())
-                        } else {
-                            Recipient::InternalAccount(
-                                *output.account(),
-                                Note::Sapling(output.note().clone()),
-                            )
+                    TransferType::Outgoing => {
+                        //TODO: Recover the UA, if possible.
+                        let recipient = Recipient::Sapling(output.note().recipient());
+                        wallet::put_sent_output(
+                            wdb.conn.0,
+                            &wdb.params,
+                            *output.account(),
+                            tx_ref,
+                            output.index(),
+                            &recipient,
```

Comment:
Okay, the 3-way match now makes clear that there's an asymmetry currently between `TransferType::Outgoing` and `TransferType::Incoming` when it comes to cross-account transfers. Both will put an entry into `sent_notes`, but only `Incoming` fills in both the funding and receiving accounts; `Outgoing` just sets the funding account but not the receiving account.

In general I think we are "probably okay" here, because we scan with IVKs before OVKs, so in the case of a cross-account transfer we will see `Incoming` here instead of `Outgoing`. However, in the future case of actually supporting imported accounts, there will be a bug here depending on the order in which accounts are imported. `wallet::put_sent_output` overwrites the contents of the `to_account_id` column, so if the receiving account gets handled before the sending account, then the collected data will be erased.

This is not simple to recover from: we cannot fix the data inconsistency in a migration if we don't have the raw transaction (to re-decrypt), and we are not guaranteed to have that if the caller is not running enhancement. So I think there are two possible options here:
- Track the receiving account in `TransferType::Outgoing`, as I had previously suggested.
- Change `wallet::put_sent_output` to not overwrite known `to_account_id` values.

Reply:
I prefer the latter solution, because in general for outgoing transfers we don't know anything about the recipient (and it would be a pain to recover that in the send flow), whereas our usual upsert functionality is to not overwrite known data with nulls.

Reply:
Within the lifetime of an a wallet database, I can't see how account IDs could ever become unstable; if you swapped them around, you'd have to update all the foreign keys everywhere and this is no exception.